### PR TITLE
HDX-9731 Add a patches table

### PR DESCRIPTION
**Pull request by IanHopkinson**
This PR implements the [HDX-9731](https://humanitarian.atlassian.net/browse/HDX-9731) ticket to add a patches table to the database. A database view is not required for this table. 

A test is added including a data fixture.

There are some small updates to documentation, and a minor change to the Ruff configuration to ignore missing docstrings.

I had a git mishap and pushed some commits direct to main, this commits have been cherry picked to the current branch to make the PR coherent.

[HDX-9731]: https://humanitarian.atlassian.net/browse/HDX-9731?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .config/ruff.toml [hunk 1]
```diff
@@ -9,7 +9,10 @@ select = [
   "I"    # isort
 ]
 ignore = [
-    "E501" # Line too long
+    "E501", # Line too long
+    "D100", # missing-module-docstring
+    "D101", # missing-class-docstring
+    "D103" # missing-function-docstring
```
Comment (turnerm) [inline .config/ruff.toml:15]: Why turn this off, do you not like writing docstrings Ian? :D

File: src/hapi_schema/db_patches.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    patch_path: Mapped[str] = mapped_column(String(128), nullable=False)
+    permanent_download_url: Mapped[str] = mapped_column(
+        String(1024), nullable=False, unique=True
+    )
+    state: Mapped[str] = mapped_column(String(16), nullable=False)
+    # discovered -> it was found in the patch repo
+    # executed -> the patch was executed successfully
+    # failed -> HWA tried to execute the patch but it failed (either pre-conditions
+    #     were not met OR the transaction failed and was rolled back)
+    # canceled -> it was marked as canceled in the patch repo
```
Comment (turnerm) [inline src/hapi_schema/db_patches.py:33]: Would it make sense to make this a constraint?

File: tests/test_patches.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        state VARCHAR(16) NOT NULL,
+        PRIMARY KEY (id),
+        UNIQUE (commit_hash),
+        UNIQUE (permanent_download_url)
+)
+"""
+    captured_sql, _ = capfd.readouterr()
+    assert sql_table_creation_code.translate(
+        str.maketrans("", "", string.whitespace)
+    ) in captured_sql.translate(str.maketrans("", "", string.whitespace))
```
Comment (turnerm) [inline tests/test_patches.py:28]: This is nice, we should probably test all the other tables this way

**Review by turnerm — APPROVED** [review]
LGTM, don't forget to update the changelog

File: src/hapi_schema/db_patches.py (hunk 2 above)
Comment (IanHopkinson) [reply src/hapi_schema/db_patches.py:33]: I'll check out how to do this - I wondered about making the state values a separate table but applying a constraint is cleaner.

File: .config/ruff.toml (hunk 1 above)
Comment (IanHopkinson) [reply .config/ruff.toml:15]: The last project I worked on they fanatically excluded comments and docstrings from code! In any case most the classes in this project lack docstrings.

Typically on line length I set something longer than the default (usually 100 characters) and then comply with it pretty rigorously.

File: tests/test_patches.py (hunk 3 above)
Comment (IanHopkinson) [reply tests/test_patches.py:28]: I noticed all the other tests were for views, so I did my own thing also I wasn't actually convinced sqlalchemy was making any SQL!

File: src/hapi_schema/db_patches.py (hunk 2 above)
Comment (IanHopkinson) [reply src/hapi_schema/db_patches.py]: I did this using a constraint: ` CheckConstraint(
            "state ='discovered' or state = 'executed' or state = 'failed' or state = 'canceled'",
            name="state_constraint",
        )` 
The in memory Sqlite database for testing does not allow subqueries in constraints excluding an "in" method and although postgres supports enums this feature seems to be dialect dependent

Comment (IanHopkinson) [reply src/hapi_schema/db_patches.py]: Following a comment from @alexandru-m-g I made this an Enum, it isn't supported by the in-memory SQlite test database but I think I prefer it over the CheckConstraint route.

**Comment by IanHopkinson** [comment]
@turnerm I added a changelog entry

**Review by alexandru-m-g — APPROVED** [review]

File: .config/ruff.toml (hunk 1 above)
Comment (turnerm) [reply .config/ruff.toml]: Hm good point that it's not catching the missed class docstrings, I will investigate

File: tests/sample_data/data_patches.py
[the remark this answers is not in the document]
Comment (turnerm) [reply tests/sample_data/data_patches.py:1]: You're probably right here Ian!

**Pull request merged**